gp env to respect workspace env vars

akosyakov opened the pull request:
## Description
<!-- Describe your changes in detail -->

`gp env` reflects workspace env vars, not only user env vars, i.e. as well coming from projects and context URL.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
fix #12208

## How to test
<!-- Provide steps to test this PR -->

- [x] Setup different project and user envs, for instance fork https://github.com/akosyakov/parcel-demo/tree/tes_gp_env
- [x] Test that `gp env` reflects visible project and context env vars.
- [x] Test that starting/restart prebuild picks up all project env vars, but not for user.
- [x] Test that starting/restart workspace picks up user, visible project and context env vars.

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
`gp env` reflects workspace env vars, not only user env vars, i.e. as well coming from projects and context URL.
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

akosyakov commented on components/server/src/workspace/workspace-starter.ts:
```diff
@@ -1580,6 +1572,22 @@ export class WorkspaceStarter {
+            // we don't have an explicit guard for a project,
+            // but we need to ensure that a workspace has access only to the associated project
+            // we assume here that a workspace don't have a way to get other project IDs besides one which we pass
+            // additionally we should be careful not exposing team specific function scopes
```
I actually clients should be able to construct such scope, so should we expose team id to the workspace as well? If we want only project id then we have to introduce a project guard.

akosyakov replied:
cc @geropl wdyt about it?

geropl replied:
I don't really understand why we need this scope in the first place. I think it's worth discussing in sync.

geropl commented on components/gitpod-protocol/go/gitpod-service.go:
```diff
@@ -2116,6 +2142,11 @@ type UpdateUserStorageResourceOptions struct {
 
+// GetWorkspaceEnvVarsOptions is the GetWorkspaceEnvVarsOptions message type
+type GetWorkspaceEnvVarsOptions struct {
+	ProjectID string `json:"projectId,omitempty"`
```
Why not parse WorkspaceID here?

akosyakov replied:
We can pass but a client still need to construct a scope which would allow to pass project guard?

geropl replied:
[Related](https://github.com/gitpod-io/gitpod/pull/16414#discussion_r1107078707)?

geropl commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1837,7 +1892,36 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
+        const result: EnvVarWithValue[] = [];
+        const { workspaceEnvVars } = await this.resolveWorkspaceEnvVars(user, {
+            repository: undefined, // filter below with access guard
+            projectId: options.projectId,
```
Why not workspace.projectId? You have to ensure that anyway, otherwise you're attaching the permission check to user-controlled data? :thinking:

akosyakov commented on components/server/src/workspace/env-var-service.ts:
```diff
@@ -0,0 +1,100 @@
+                workspace.context.repository.owner,
+                workspace.context.repository.name,
+            );
+        }
```
I wonder why do we return all user envs otherwise? Should not we have non user env vars outside of commit context?

svenefftinge replied:
The commit context gives us the repository name which we use for the filtering. I.e. without the commit context we don't know the repository.

akosyakov replied:
yeah, i mean if context is not commit then user envs should be empty? it seems right now it is everything on main?

akosyakov commented on components/gitpod-cli/cmd/env.go:
```diff
@@ -97,14 +98,15 @@ func connectToServer(ctx context.Context) (*connectToServerResult, error) {
 		Kind: "gitpod",
 		Scope: []string{
-			"function:getEnvVars",
+			"function:getWorkspaceEnvVars",
```
TODO: actually compatibility check should happen here, there won't be token otherwise

akosyakov commented:
@svenefftinge @geropl I have not tested yet, going after the lunch, but I've refactored it that now there is `EnvVarService.resolve` which hides the merge logic of env vars from different sources given user (optional) or workspace. All code in Server is using it, i.e. for create/start workspace, run/rerun prebuilds, and resolve env vars for the running workspace. I also added tests for the merge logic. Could you have a look please? 🙏

akosyakov commented on components/server/src/workspace/env-var-service.ts:
```diff
@@ -0,0 +1,92 @@
+        };
+
+        // 1. first merge user envs
+        let userEnvVars = user ? await this.userDB.getEnvVars(user.id) : [];
```
Should we have prebuild guard here, instead of relying on passing `user` as null? i.e. a user can be passed always but will be ignored if workspace type is prebuild

svenefftinge replied:
Since we are looking up the project from the workspace, we could also look up the user from it. And then also considering the workspace.type. So just passing in the workspace.

akosyakov replied:
Just to check you mean to use `workspace.ownerId`

svenefftinge commented on components/server/src/workspace/env-var-service.ts:
```diff
@@ -0,0 +1,92 @@
+    project: ProjectEnvVar[];
+    context: EnvVarWithValue[];
+    // merged workspace env vars
+    workspace: EnvVar[];
```
why do we expose the individual layers here? Isn't just returning the workspace array enough?

akosyakov replied:
Projects are used a bit differently to figure image build auth value. I added other to assert in tests. I can reduce it to workspace and projects env vars only and don't assert other layers in tests?

akosyakov commented on components/server/src/workspace/workspace-starter.ts:
```diff
@@ -686,9 +677,9 @@ export class WorkspaceStarter {
+    protected async getAdditionalImageAuth(envVars: ResolvedEnvVars): Promise<Map<string, string>> {
         const res = new Map<string, string>();
-        const imageAuth = projectEnvVars.find((e) => e.name === "GITPOD_IMAGE_AUTH");
+        const imageAuth = envVars.project.find((e) => e.name === "GITPOD_IMAGE_AUTH");
```
@svenefftinge this is the major reason why we need to return all, although we could change that this method itself fetches a required env var.

svenefftinge replied:
I wonder why this env var needs to be set on project-level and not anyhow.

akosyakov replied:
@svenefftinge I removed other layers, but kept this one for now. I think someone else can address it later if needed. Looking at the original PR and docs it seems to be that intent was to allow it only on project env vars.

akosyakov commented:
/werft run

akosyakov commented on components/server/src/workspace/workspace-starter.ts:
```diff
@@ -1294,47 +1285,17 @@ export class WorkspaceStarter {
-
-        const envvars: EnvironmentVariable[] = [];
-
         // TODO(cw): for the time being we're still pushing the env vars as we did before.
```
@csweichel here an idea was that we don't pass it anymore, but supervisor fetches on start after getting the token?

I noticed that ws-manager has a filter for workspace env vars: https://github.com/gitpod-io/gitpod/blob/dda547c7a2803b36457998a1440bc9485e9be439/components/ws-manager-mk2/controllers/create.go#L605-L623

Should not we rather filter them here? that this behaviour is consistent with `gp env`?

akosyakov replied:
We have now 2 kind of env vars on spec, one configured by user for a workspace (user, project, context), and then system env vars (which are not filtered).

Let's move `GITPOD_WORKSPACE_CONTEXT` and so on to system env vars, and then filter the rest in `EnvVarService`?

akosyakov replied:
It is complicated because of protected secrets. If supervisor would fetch env vars instead it would be possible to simplify since we don't leave env vars on k8s spec then. I will leave it out for another PR.

akosyakov commented:
/werft run

akosyakov commented:
Output of `gp env` with this PR:
- start after prebuild
<img width="350" alt="Screenshot 2023-02-17 at 11 49 47" src="https://user-images.githubusercontent.com/3082655/219626101-e69221d0-fc07-4712-8f95-48869a1e30bc.png">

- restart
<img width="750" alt="Screenshot 2023-02-17 at 11 51 57" src="https://user-images.githubusercontent.com/3082655/219626434-840d4b9a-994a-4791-ba13-a9ebc158d4b4.png">

Output of `env` reflects the same.

akosyakov commented:
/hold

just want @geropl to double check, can someone from @gitpod-io/engineering-ide to review 🙏

geropl reviewed (approved):
LGTM

akosyakov commented:
/unhold
